src/CSnakes.Runtime/Python/Pack.cs

```diff
@@ -10,46 +15,233 @@ namespace CSnakes.Runtime.Python;
+
+        public static RentedArray<nint> RentHandles(int length, ref RentalState rental) => new(Handles, length, ref rental);
+        public static RentedArray<PyObjectMarshaller.ManagedToUnmanagedIn> RentMarshallers(int length, ref RentalState returned) => new(Marshallers, length, ref returned);
+    }
+
+    private static nint CreateListOrTuple<TBuilder>(Span<PyObject> items)
+        where TBuilder : IListOrTupleBuilder
+    {
+        // Allocate initial space for the handles and marshallers on the stack.
+        // If the number of items exceeds the stack space, allocate and spill
+        // the rest into an array on the heap.
+
+        const int stackSpillThreshold = FixedArrayLength;
+        var spillLength = Math.Max(0, items.Length - stackSpillThreshold);
+
+        Span<nint> initialHandles = stackalloc nint[Math.Min(stackSpillThreshold, items.Length)];
+        var spilledHandlesRental = new RentalState();
+        using var spilledHandles = ArrayPools.RentHandles(spillLength, ref spilledHandlesRental);
+
+        var initialMarshallers = new ArrayOf8<PyObjectMarshaller.ManagedToUnmanagedIn>();
+        var spilledMarshallersRental = new RentalState();
+        using var spilledMarshallers = ArrayPools.RentMarshallers(spillLength, ref spilledMarshallersRental);
+
+        scoped var uninitializedMarshallers = MemoryMarshal.CreateSpan(ref Unsafe.As<ArrayOf8<PyObjectMarshaller.ManagedToUnmanagedIn>, PyObjectMarshaller.ManagedToUnmanagedIn>(ref initialMarshallers), stackSpillThreshold);
+        var uninitializedHandles = initialHandles;
+
+        // The following loop initializes the marshallers and handles for each
+        // item in the input span. It is assumed that no exceptions are thrown
+        // during this loop. The marshallers are freed in the finally block of
+        // the actual list/tuple initialization.
 
         foreach (var item in items)
         {
-            int result = CPythonAPI.PyList_Append(pyList, item);
-            if (result == -1)
+            PyObjectMarshaller.ManagedToUnmanagedIn m = default;
+            m.FromManaged(item);
+
+            if (uninitializedMarshallers.IsEmpty && spilledMarshallers is { Length: > 0 } sms)
+            {
+                uninitializedMarshallers = sms;
+            }
+
+            uninitializedMarshallers[0] = m;
+            uninitializedMarshallers = uninitializedMarshallers[1..];
+
+            if (uninitializedHandles.IsEmpty && spilledHandles is { Length: > 0 } shs)
+            {
+                uninitializedHandles = shs;
+            }
+
+            uninitializedHandles[0] = m.ToUnmanaged();
+            uninitializedHandles = uninitializedHandles[1..];
+        }
+
+        nint obj = 0;
+
+        try
+        {
+            obj = TBuilder.New(items.Length);
+            SetItems(spilledHandles, SetItems(initialHandles, 0));
+
+            return obj;
+
+            int SetItems(Span<nint> handles, int i)
```

Comment:
I would pass all state in and mark this `static`. Capturing local functions are ill-advised.

Reply:
Resolving as this is moot since it's gone with a1c1625a57684cc75b434cfcbd9b604ecebc0e57.

Reply:
> I'd be curious to learn and understand why. If this is about potentially inducing a heap-allocation due to a closure then I don't believe it happens in this particular case. The compiler generated code for the local function gets lowered to this:

Your assumption is correct - inadvertant heap-allocation coupled with additional code-gen. The biggest issue with captured state is the opaque nature of what is generated. The C# compiler doesn't state what is generates and forcing people to look at the generated IL isn't something that should be expected. In this case it also breeds additional capture when they just aren't needed.